test/test_rack_server.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require "test/unit"
+require "rack/test"
+require "xmlrpc/server"
+require 'xmlrpc/create'
+require 'xmlrpc/parser'
+
+class TestRack < Test::Unit::TestCase
+  include Rack::Test::Methods
+
+  def app
+    s = XMLRPC::RackServer.new
+
+    s.add_handler("test.add") do |a,b|
+      a + b
+    end
+
+    s.add_handler("test.div") do |a,b|
+      if b == 0
+        raise XMLRPC::FaultException.new(1, "division by zero")
+      else
+        a / b
+      end
+    end
+
+    s.set_default_handler do |name, *args|
+      raise XMLRPC::FaultException.new(-99, "Method #{name} missing" +
+            " or wrong number of parameters!")
+    end
+
+    s.add_introspection
+
+    s
+  end
+
+  def test_successful_call
+    assert_equal([true, 9], call("test.add", 4, 5))
+  end
+
+  def test_fault_exception
+    assert_equal([false, XMLRPC::FaultException.new(1, "division by zero")], call("test.div", 1, 0))
+  end
+
+  def test_introspection
+    methods = ["test.add", "test.div", "system.listMethods", "system.methodSignature", "system.methodHelp"]
+    assert_equal([true, methods], call("system.listMethods"))
+  end
+
+  def test_missing_handler
+    assert_equal([false, XMLRPC::FaultException.new(-99, "Method test.nonexisting missing or wrong number of parameters!")], call("test.nonexisting"))
+  end
+
+  def test_wrong_number_of_arguments
+    assert_equal([false, XMLRPC::FaultException.new(-99, "Method test.add missing or wrong number of parameters!")], call("test.add", 1, 2, 3))
+  end
+
+  def test_multibyte_characters
+    assert_equal([true, "あいうえおかきくけこ"], call('test.add', "あいうえお", "かきくけこ"))
+  end
+
+  def call(methodname, *args)
+    create = XMLRPC::Create.new(XMLRPC::Config.default_writer.new)
+    parser = XMLRPC::Config.default_parser.new
+
+    request = create.methodCall(methodname, *args)
+    post("/", request, 'CONTENT_TYPE' => 'text/xml')
+    assert(last_response.ok?)
```

Comment:
Could you use power assert for easy to debug? BTW, when does `ok?` return `false`?

```suggestion
    assert do
      last_response.ok?
    end
```

Reply:
https://github.com/rack/rack/blob/main/lib/rack/response.rb#L189

`ok?` checks the HTTP status code for 200. This is mostly relevant for the exceptions, where XMLRPC will still return a HTTP OK (200) status, as opposed to using the HTTP status to signal an error like most REST-based APIs do.

Reply:
This results in output like this:
```        last_response.ok?
        |             |
        |             false
        #<Rack::MockResponse:0x00007fbc70b25aa0 @original_headers={"Content-type"=>"text/xml; charset=utf-8"}, @errors="", @status=201, @headers={"content-type"=>"text/xml; charset=utf-8", "content-length"=>"296"}, @writer=#<Method: Rack::MockResponse(Rack::Response::Helpers)#append(chunk) /home/herwin/xmlrpc/vendor/bundle/ruby/3.1.0/gems/rack-3.1.13/lib/rack/response.rb:359>, @block=nil, @body=["<?xml version=\"1.0\" ?><methodResponse><fault><value><struct><member><name>faultCode</name><value><i4>-99</i4></value></member><member><name>faultString</name><value><string>Method test.add missing or wrong number of parameters!</string></value></member></struct></value></fault></methodResponse>\n"], @buffered=true, @length=296, @cookies={}>
/home/herwin/xmlrpc/test/test_rack_server.rb:73:in `call'
/home/herwin/xmlrpc/test/test_rack_server.rb:59:in `test_wrong_number_of_arguments'
     56: 
     57:   def test_wrong_number_of_arguments
     58:     assert_equal([false, XMLRPC::FaultException.new(-99, "Method test.add missing or wrong number of parameters!")],
  => 59:                  call("test.add", 1, 2, 3))
     60:   end
```
I don't think this is the most readable output, I think the better option would be something like:
```ruby
assert(last_response.ok?, "Expected HTTP status code 200, got #{last_response.status} instead")
```

Reply:
OK. It's also better than `assert(last_response.ok?)`.